library/ssl_cache.c

```diff
@@ -140,117 +139,116 @@ static int ssl_cache_pick_writing_slot( mbedtls_ssl_cache_context *cache,
     cur = cache->chain;
     prv = NULL;
 
+    /* Check 1: Is there already an entry with the given session ID?
+     *
+     * If yes, overwrite it.
+     *
+     * If not, `count` will hold the size of the session cache
+     * at the end of this loop, and `prv` will point to the last
+     * entry, both of which will be used later. */
+
     while( cur != NULL )
     {
         count++;
 
-#if defined(MBEDTLS_HAVE_TIME)
-        if( cache->timeout != 0 &&
-            (int) ( t - cur->timestamp ) > cache->timeout )
-        {
-            cur->timestamp = t;
-            break; /* expired, reuse this slot, update timestamp */
-        }
-#endif
-
         if( session_id_len == cur->session_id_len &&
             memcmp( session_id, cur->session_id, cur->session_id_len ) == 0 )
         {
-            break; /* client reconnected, keep timestamp for session id */
+            goto found;
         }
 
+        prv = cur;
+        cur = cur->next;
+    }
+
+    /* Check 2: Is there an outdated entry in the cache?
+     *
+     * If so, overwrite it.
+     *
+     * If not, remember the oldest entry in `old` for later.
+     */
+
 #if defined(MBEDTLS_HAVE_TIME)
+    while( cur != NULL )
```

Comment:
Shouldn't we reset `cur` to `cache->chain` before we enter this loop?

Reply:
Gosh. Absolutely. Thanks for spotting.

Reply:
We should have tests exercising this. As this is an existing test gap, though, I'd prefer to add this to https://github.com/ARMmbed/mbedtls/issues/4508 and handle it at a later point.

Reply:
Yeah, I worked on some untested code recently and managed to introduce a mistake that breezed through all the compilers and tests. My luck was that MSVC started wondering if that is really what I want to do 😄 

I agree, this is best done in #4508 and not in this PR.